style: adopt black

CaselIT opened the pull request:
Just to get the ball rolling.

Tried to run black ~~20.8b1~~ 21.5b1 with the configuration
```toml
[tool.black]
    target-version = ["py35"]
    skip-string-normalization = true
    line-length = 88
```
The `skip-string-normalization` avoids the conversion `'` to `"` if we want to keep the current flake8 rules

TODO:
- [x] re-run with line length 88
- [x] cleanup string concatenation in the same line (like `'foo bar ' 'baz foobar'`)
- [x] review where we need to add / remove tailing commas to improve the styling
- [ ] update the documentation with the changes done to the example folder

Closes #1453

CaselIT commented on examples/look/look/images.py:
```diff
@@ -8,18 +8,11 @@
-                }
-            ]
-        }
+        doc = {'images': [{'href': '/images/1eaf6ef1-7f2d-4ecc-a8d5-6e8adba7cc0e.png'}]}
```
NOTE: this could be improved a bit: placing a comma `,` at the end of the end of a definition makes black use new lines, so we can force expanded formatting for dict/string/argument lists, so that the formatted version is closer to the existing one. It would become:

```py
        doc = {
            'images': [
                {
                    'href': '/images/1eaf6ef1-7f2d-4ecc-a8d5-6e8adba7cc0e.png',
                },
            ],
        }
```

vytas7 replied:
IIRC this look stuff used to be copypasted, so we'll need to sync, or use "literal include" like I did in the ASGI version (thanks @CaselIT).

CaselIT replied:
ok let's add some points to do

vytas7 commented on examples/things_advanced.py:
```diff
@@ -45,48 +43,53 @@ def __call__(self, req, resp, engine):
         if token is None:
-            description = ('Please provide an auth token '
-                           'as part of the request.')
+            description = 'Please provide an auth token ' 'as part of the request.'
```
These changes look odd. We would probably need to carefully review them manually if we want to keep the same high standard of Falcon's source code.

CaselIT replied:
reaching for `' '` finds 92 results. I think it would be manageable

vytas7 replied:
Wondering if `black` is making any progress in this area? :thinking: 
IIRC there were discussions to handle this out of the box, but I haven't checked what the status is.

CaselIT replied:
Not much. It now remosed leading and trailing spaces in triple comments where it makes sense.

But thsi caae needs to be handled by hand.

vytas7 replied:
I think the intention is to handle all strings: https://github.com/psf/black/issues/2188

But it seems there are still some issues with it.

CaselIT replied:
ah, sqlalchemy makes black file, ahah :)

CaselIT replied:
tried running it. not particularly convinced:

see the diff https://gist.github.com/CaselIT/2e509ce7b7425700a87f190669baeb1c

kgriffs commented:
I am going to delay this for post 3.0 but I do think we should move in this direction. Let's just make sure we have as part of the process a review to tweak the occasional oddity.

CaselIT commented:
> 
> 
> I am going to delay this for post 3.0 but I do think we should move in this direction. Let's just make sure we have as part of the process a review to tweak the occasional oddity.

yes sure, I was not targeting this release, it was just to start looking into it

CaselIT commented:
I've updated the change with the latest black.

kgriffs commented:
The default of 88 is fine, and avoids excessive line wrapping esp. for deeply nested control structures. If we increased the limit much further I would argue for a separate standard for keeping comments shorter (even if we have to write our own checker for it).

Regardless, I'll take a look through the diff soon and see how things look now.

CaselIT commented:
I'll re-reun with the 88 line.

If you are not familiar with black, note that you can tell it to split things in multiple lines by setting a comma at the end.
Like
```py
def foo(x,):
  pass
# becomes
def foo(
  x,
):
  pass
````
So we may need to add / remove some to get the desired formatting.

https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html?highlight=comma#the-magic-trailing-comma

vytas7 commented:
I'm a bit orthodox when it comes to line length, 100+ sounds a bit too long.

But I'm fine with `black`'s 88, it does sound like a good compromise, and... if we're going to use `black`, we don't need to care about manually keeping lines at 79 or 88 anyway, that's what `black` is for.

CaselIT commented:
Ok, I initially went with 100 both because we do have some lines that are on the long side in the source and it's what I usually use.

I'll re-run with the default 88 and update this.

CaselIT commented:
Updated with 88. From the change lines it seems that our average line was a bit longer, since this version adds a 1000 new lines.

For reference the previous with length 100 was more or less a wash regarding line count

kgriffs reviewed (changes requested):
Comments inline. I am still favorable to adopting black modulo a few tweaks. Also, note that f we miss any strange formatting issues before merging this, we can always fix those up over time, as we discover them in the course of maintaining the project.

CaselIT commented on falcon/app.py:
```diff
@@ -1087,7 +1109,8 @@ class API(App):
-    @deprecation.deprecated('API class may be removed in a future release, '
-                            'use falcon.App instead.')
+    @deprecation.deprecated(
+        'API class may be removed in a future release, ' 'use falcon.App instead.'
```
sure I've yet to make this pass

CaselIT commented on falcon/media/validators/jsonschema.py:
```diff
@@ -106,22 +106,19 @@ def wrapper(self, req, resp, *args, **kwargs):
                 raise falcon.HTTPBadRequest(
-                    title='Request data failed validation',
-                    description=e.message
+                    title='Request data failed validation', description=e.message
```
well, it depends if it fits or not on the line (without tailing comma that is).

it's usually ok, and for the cases where it's better as single-lines adding a comma it's easy. I would guess that we will adjust the formatting a bit over time, to accommodate for these when we modify a file that's formatted funny :)

CaselIT commented:
I'll wait for https://github.com/falconry/falcon/pull/1709 before rebasing and polishing this

kgriffs commented on examples/look/tests/test_app.py:
```diff
@@ -48,9 +42,7 @@ def test_post_image(client, mock_store):
-        '/images',
-        body=b'some-fake-bytes',
-        headers={'content-type': image_content_type}
+        '/images', body=b'some-fake-bytes', headers={'content-type': image_content_type}
```
> always add magic commas to be safe

...for code samples that will be included in the docs, that is.

CaselIT commented on docs/user/quickstart.rst:
```diff
@@ -17,46 +17,8 @@ started writing an app.
-                    # Serve until process is killed
-                    httpd.serve_forever()
+        .. literalinclude:: ../../examples/things.py
+            :language: python
```
`literalinclude` ftw

CaselIT commented:
@kgriffs @vytas7 I've updated the docs to use the new formatting, bug I'm not sure I updated all of them.

I've also made a quick pass to the files to check that no funny formatting was added. For this I think we can correct eventual leftover in the future

vytas7 commented on setup.cfg:
```diff
@@ -85,7 +85,7 @@ filterwarnings =
 exclude = .ecosystem,.eggs,.git,.tox,.venv,build,dist,docs,examples,falcon/bench/nuts
-ignore = F403,W504
+extend-ignore = F403,W504,E203,I202
 max-line-length = 99
```
Shouldn't we now decrease this to 88?

CaselIT replied:
Yes, I guess so

CaselIT replied:
Note that we have quite a bit of comment/documentation that are over 88 chars

vytas7 commented on tests/asgi/test_boundedstream_asgi.py:
```diff
@@ -154,14 +164,18 @@ async def test_iteration():
     falcon.async_to_sync(test_iteration)
 
 
-@pytest.mark.parametrize('body', [
```
Ah, these test fixtures got a bit ugly, but I guess we'll have to live with that.
Unless there is some magic opening parenthesis out there :laughing:

CaselIT replied:
I actually don't mind them, but it may be that I'm used to them

vytas7 replied:
Me neither, it's just a massive refactoring of something that was aesthetically OK/better before.

CaselIT replied:
I guess it's inevitable when introducing a formatter that a large number of pointless changed are done.

kgriffs reviewed (dismissed):
💥

vytas7 commented on falcon/routing/compiled.py:
```diff
@@ -224,13 +224,19 @@ def insert(nodes, path_index=0):
+                    msg = (
+                        textwrap.dedent(
+                            """
                         The URI template for this route is inconsistent or conflicts with another
```
Shouldn't the text be moved further left to match the triple quotes?

CaselIT replied:
I'll check if it can be tweaked

vytas7 commented on falcon/util/uri.py:
```diff
@@ -25,21 +25,19 @@
-               'abcdefghijklmnopqrstuvwxyz'
-               '0123456789'
-               '-._~')
+_UNRESERVED = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~'
```
Does `black` insist on joining these, or is possible to leave the separation as-is?

CaselIT replied:
yes, since they fit in the same line it will do `'abcdefghijklmnopqrstuvwxyz' '0123456789' '-._~'` I joined them together.

we could mark the section as format off, or use a tuple and join them. But I think it's fine to have it as one line.

Also maybe we could use something linke `string.ascii_letters + string.digits + '-._~'` ?

vytas7 replied:
This one is not critical for me, I just thought that the previous indentation was intentional.

CaselIT replied:
> I just thought that the previous indentation was intentional.

that's most likely true, but I think it's still understandable that's ascii + digits + some special chars

kgriffs replied:
string.ascii_letters + string.digits + '-._~' sounds like a good way to make it a little more self-documenting, but not critical to change it.

CaselIT replied:
The only thing to consider is that with the concatenation form the resulting string may not be interned by python (iirc the string literal are).

If we want to change this a follow on pr is probably best

vytas7 commented on falcon/vendor/mimeparse/mimeparse.py:
```diff
@@ -31,8 +31,7 @@ def parse_mime_type(mime_type):
     if not type_parts or len(type_parts) > 2:
-        raise MimeTypeParseException(
-            "Can't parse type \"{}\"".format(full_type))
+        raise MimeTypeParseException("Can't parse type \"{}\"".format(full_type))
```
Maybe we should exclude vendored modules from blackening?

It might make it more cumbersome to backport changes/sync with upstream, even though any changes are looking unlikely at this point.

CaselIT replied:
oh good point. let me try to exclude them.

vytas7 reviewed (commented):
This is getting almost ready :+1: 

I just had a handful of open questions (commented inline).

And a more general question: should we try updating Cython code for the delimited stream reader etc? When I wrote it, it was loosely based on the Python version, with changes specific to Cython. Right now, that parity would be disturbed a bit. OTOH mangling that code now after the fact is just a (an unnecessary?) risk to introduce regressions.

CaselIT commented:
> And a more general question: should we try updating Cython code for the delimited stream reader etc? When I wrote it, it was loosely based on the Python version, with changes specific to Cython.

black does not support cython, so the formatting would be manual.

Personally I think we could leave it as is or update as needed as a follow on change (maybe in the same issue as the one setting flake line length to 88)

kgriffs reviewed (approved):
👍

vytas7 reviewed (approved):
Here we go then...

CaselIT commented:
I've opened https://github.com/falconry/falcon/issues/1926 to do the follow up on this change.

Please add there if there were other deferred points from this change. I'll try to get to it this we
